src/Testcontainers/Builders/AbstractBuilder`2.cs

```diff
@@ -0,0 +1,137 @@
+﻿namespace DotNet.Testcontainers.Builders
+{
+  using System;
+  using System.Collections.Generic;
+  using DotNet.Testcontainers.Clients;
+  using DotNet.Testcontainers.Configurations;
+  using DotNet.Testcontainers.Containers;
+  using JetBrains.Annotations;
+
+  /// <summary>
+  /// An abstract fluent Docker resource builder.
+  /// </summary>
+  /// <typeparam name="TBuilderEntity">The builder entity.</typeparam>
+  /// <typeparam name="TResourceEntity">The resource entity.</typeparam>
+  /// <typeparam name="TConfigurationEntity">The configuration entity.</typeparam>
+  [PublicAPI]
+  public abstract class AbstractBuilder<TBuilderEntity, TResourceEntity, TConfigurationEntity> : IAbstractBuilder<TBuilderEntity, TResourceEntity>
+    where TBuilderEntity : IAbstractBuilder<TBuilderEntity, TResourceEntity>
+    where TConfigurationEntity : IResourceConfiguration
+  {
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AbstractBuilder{TBuilderEntity, TResourceEntity, TConfigurationEntity}" /> class.
+    /// </summary>
+    /// <param name="dockerResourceConfiguration">The Docker resource configuration.</param>
+    protected AbstractBuilder(TConfigurationEntity dockerResourceConfiguration)
+    {
+      _ = TestcontainersSettings.SettingsInitialized.WaitOne(TimeSpan.FromSeconds(5));
+      this.DockerResourceConfiguration = dockerResourceConfiguration;
+    }
+
+    /// <summary>
+    /// Gets the Docker resource configuration.
+    /// </summary>
+    protected virtual TConfigurationEntity DockerResourceConfiguration { get; }
+
+    /// <inheritdoc cref="IAbstractBuilder{TBuilderEntity, TContainerEntity}" />
+    public TBuilderEntity WithDockerEndpoint(string endpoint)
```

Comment:
> `IDockerEndpointAuthenticationConfiguration` seems to represent such an abstraction.

The overloaded method `WithDockerEndpoint(IDockerEndpointAuthenticationConfiguration)` allows developers to pass a "custom" Docker client implementation to Testcontainers for .NET.

> or is it easier if the implementation internally takes care of reusing a client if the same endpoint configuration is given?

.NET caches the client object internal (it is not created immediately). The [auto discovery value](https://github.com/testcontainers/testcontainers-dotnet/blob/c126fb91f7e7872a72b9086f0dcca76b03b54507/src/Testcontainers/Builders/AbstractBuilder%602.cs#L109) is set as default endpoint for each configuration.

> And for which users is this API relevant in the first place?

I know developers are using it to configure the HTTP client.

Reply:
We will resolve this comment by extending the method documentation to bring an emphasis on this being a method not normally called by users (since they should rely on autoconfig).

Reply:
I've added

> Testcontainers automatically discovers the Docker environment and applies the configuration.
> It is recommended to use the auto discovery. Only certain edge cases require this enhanced API.

to the XML (code) documentation.